### ISSUE TITLE
AMP is not working with Material-UI

### ISSUE THREAD

**Issue by dmdb** [zeit/next.js#9041]
# Examples bug report

## Example name

https://github.com/mui-org/material-ui/tree/master/examples/nextjs

## Describe the bug

If you turn on AMP-first on a page, for example on about page:
![image](https://user-images.githubusercontent.com/427746/66646058-4377b180-ec4f-11e9-9630-a09b1c82d148.png)
You will get error
![image](https://user-images.githubusercontent.com/427746/66646096-64400700-ec4f-11e9-9ea1-2927a55335cb.png)

## To Reproduce

1. Follow instructions https://github.com/mui-org/material-ui/tree/master/examples/nextjs
2. Enable AMP on page
3. Open it

## Expected behavior

AMP is working fine with Material UI

## System information

- OS: Ubuntu 18.04
- Chrome
- Next.js 9.1.1, MaterialUI 4.5.0

## Additional context

It maybe problem with material-ui example or other styling integrations. Here is some similar as I think issues:
https://github.com/zeit/next.js/issues/7856
https://github.com/mui-org/material-ui/issues/17784

**Comment by Janpot** [comment]
@dmdb FYI, in case you want to pick this up, an effort to fix this was started in https://github.com/zeit/next.js/pull/7954, I'm not 100% sure where this should be headed, but I guess the example should be updated to use

```js
        styles: [
          ...React.Children.toArray(initialProps.styles),
          ...sheet.getStyleElement(),
        ],
```

to make it work and compliant with the types.

**Comment by dmdb** [comment]
@Janpot Unfortunately I'am not quite confident to pick it, I've just monkey-patched it in our project (thanks to @tomevans18) with
![image](https://user-images.githubusercontent.com/427746/66655982-4a122300-ec67-11e9-9444-1896a84e3dcd.png)
It seems works fine, but not sure why :woman_shrugging: 
Anyway it seems that it is a common problem with example and should be fixed somehow in proper, official way.

PS. Thanks for AMP support on Next.js team. It is a huge benefit

**Comment by oliviertassinari** [comment]
We have updated our demo to work, and now waiting on https://github.com/mui-org/material-ui/issues/18044 to solve the TypeScript part of it.

**Comment by Danhiler** [comment]
I'm using the Material-UI's example of Next.js and AMP, I get the following error:
**Cannot read property 'dangerouslySetInnerHTML' of undefined**
Here is my code:
**styles: [...React.Children.toArray(initialProps.styles), materialUiSheets.getStyleElement(), styledComponentSheet.getStyleElement()]**

What am I doing wrong?

**Comment by derolf** [comment]
I am trying to get nextjs with amp and material-ui to work, but the styles don’t get included in amp mode (works in non-amp mode). Is there any minimal example?

**Comment by KyruCabading** [comment]
@oliviertassinari saw this and wanted to mention that NextJS has a custom Html component within `next/document` that might be what's missing for AMP support. One of our devs noticed that new component in a NextJS MUI PR

If you look at docs here, there's a new Html component to use within _document.js
https://nextjs.org/docs/advanced-features/custom-document

And that component adds an amp prop which is required for AMP support. 
https://github.com/zeit/next.js/blob/canary/packages/next/pages/_document.tsx#L138-L165

https://amp.dev/documentation/guides-and-tutorials/learn/spec/amphtml/

**Comment by oliviertassinari** [comment]
@KyruCabading As far as I know, we have fixed this issue, it can be closed.

**Comment by timneutkens** [comment]
Opened https://github.com/mui-org/material-ui/pull/20463, good catch @KyruCabading

**Comment by oliviertassinari** [comment]
@timneutkens Thanks, we are making progress:

<img width="604" alt="Capture d’écran 2020-04-08 à 11 48 18" src="https://user-images.githubusercontent.com/3165635/78770856-7ea13600-798f-11ea-96be-0de4e314a335.png">

However, when I try to render this page

```jsx
import React from 'react';
import TextField from '@material-ui/core/TextField';

export default function BasicTextFields() {
  return (
    <div>
      <TextField id="standard-basic" label="Standard" />
    </div>
  );
}

export const config = { amp: true }
```

I get the following error:

```
Creating an optimized production build

Compiled successfully.

Amp Validation

/  error  The text inside tag 'style amp-custom (transformed)' contains 'CSS !important', which is disallowed.  https://amp.dev/documentation/guides-and-tutorials/learn/spec/amphtml#stylesheets
   error  The text inside tag 'style amp-custom (transformed)' contains 'CSS !important', which is disallowed.  https://amp.dev/documentation/guides-and-tutorials/learn/spec/amphtml#stylesheets
   error  The text inside tag 'style amp-custom (transformed)' contains 'CSS !important', which is disallowed.  https://amp.dev/documentation/guides-and-tutorials/learn/spec/amphtml#stylesheets
   error  The text inside tag 'style amp-custom (transformed)' contains 'CSS !important', which is disallowed.  https://amp.dev/documentation/guides-and-tutorials/learn/spec/amphtml#stylesheets


> Build error occurred
Error: AMP Validation caused the export to fail. https://err.sh/zeit/next.js/amp-export-validation
```

**Comment by timneutkens** [comment]
cc @sebastianbenz

**Comment by oliviertassinari** [comment]
Found it https://github.com/mui-org/material-ui/blob/cc3b03becb98902b027026081e01832b009e7507/packages/material-ui/src/InputBase/InputBase.js#L25

**Comment by oliviertassinari** [comment]
It works if I comment the `!important`. I will see if we can work around it in the core. The only concern I have left is that we rely on the React runtime to update the styles on focus. We lose this visual feedback.

<img width="579" alt="Capture d’écran 2020-04-08 à 12 05 29" src="https://user-images.githubusercontent.com/3165635/78771953-43a00200-7991-11ea-8e1d-4fc703043995.png">

**Comment by oliviertassinari** [comment]
We use `!important` to fix an issue with Edge. We have an alternative solution in https://github.com/mui-org/material-ui/issues/15183#issuecomment-480949407 but I'm not sure it's worth fixing. I propose we wait to see users' requests for it.

**Comment by balazsorban44** [comment]
This issue has been automatically locked due to no recent activity. If you are running into a similar issue, please create a new issue with the steps to reproduce. Thank you.
